### DISCUSSION TITLE
Dev circular dependency between `coder.py` and `coremodel.py`

### DISCUSSION THREAD

**Discussion by itafraze** [gaphor/gaphor#3429] [Q&A] [answered]
Hi, I just wanted to share something I came across in case you didn't know about it and it might be worth your attention.

While looking around to understand how everything works I landed onto `gaphor/codegen/coder.py` and started playing with `models/Core.gaphor` and `poetry run poe coremodel`. Once I deleted an invalid `coremodel.py`that was causing errors the following `poetry run` resulted in 

```console
$ poetry run poe coremodel
Poe => coremodel
No translations were found for language en_GB.UTF-8: [Errno 2] No translation file found for domain: 'gaphor'
Traceback (most recent call last):
...
  File "/gaphor/codegen/coder.py", line 33, in <module>
    from gaphor import UML
  File "/gaphor/UML/__init__.py", line 4, in <module>
    import gaphor.UML.umloverrides
  File "/gaphor/UML/umloverrides.py", line 12, in <module>
    from gaphor.core.modeling import Element
  File "/gaphor/core/modeling/__init__.py", line 3, in <module>
    from gaphor.core.modeling.coremodel import (
ModuleNotFoundError: No module named 'gaphor.core.modeling.coremodel'
```

P.S. I'm trying to figure out how to set the default element names for #3428. If you could point me in the right direction, I'd appreciate it.

**Comment by danyeaw** [comment]
Hi @itafraze, what module / file did you delete when you said you deleted an invalid coremodel.py?

**Reply by itafraze** [reply]
This file [gaphor/core/modeling/coremodel.py](https://github.com/gaphor/gaphor/blob/b42af224fbd89f3c85da01d894e345b6426008d8/gaphor/core/modeling/coremodel.py), which is the one generated with the coder: 

https://github.com/gaphor/gaphor/blob/b42af224fbd89f3c85da01d894e345b6426008d8/gaphor/core/modeling/coremodel.py#L1-L3

Perhaps I could have been more clear. I should have specified that I am not talking about a Python circular dependency. What I meant was that the generated file is a dependency for the generator script.

**Comment by amolenaar** [comment]
Hi @itafraze,

The dependency is intentional: we use the default functionality of Gaphor to load the model and generate a new code from a model.

We used to have a special code generator that parsed the code by itself and then tried to generate the code. This code was hard to change. By using Gaphor to generate new versions of Gaphor we basically eat our own dogfood.

Normally when I make a mistake I `git restore` the original file to get back to a working state.

Note that we also [have tests](https://github.com/gaphor/gaphor/blob/main/tests/test_models_up_to_date.py) that check if the model and generated code are in sync.

**Reply by itafraze** [reply]
Thanks for clarifying the design decision, I was curious about that. 

The `coder` function now returns an iterable, and it is used here to write to the file:

https://github.com/gaphor/gaphor/blob/6fee35e8a9413b28c3b32e5d3a1a0697dc244605/gaphor/codegen/coder.py#L106-L110

This causes any error to write an incomplete output. Hence the need to rely on git for recovery. Would you consider it an improvement to either write to a temporary file first, or to capture the entire contents of the file before flushing it to the output?

**Reply by amolenaar** [reply, marked as answer]
That will be an improvement for sure. It would also allow us to do a few basic checks (e.g. see if we can load the module at all).
